Transient failure of GroupByQueryMergeBufferTest

```
Running io.druid.query.groupby.GroupByQueryMergeBufferTest
Tests run: 20, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.865 sec <<< FAILURE! - in io.druid.query.groupby.GroupByQueryMergeBufferTest
testTripleNestedGroupBy[rtIndex](io.druid.query.groupby.GroupByQueryMergeBufferTest)  Time elapsed: 0.641 sec  <<< ERROR!
io.druid.query.QueryInterruptedException
	at io.druid.query.groupby.epinephelinae.GroupByMergingQueryRunnerV2.waitForFutureCompletion(GroupByMergingQueryRunnerV2.java:340)
	at io.druid.query.groupby.epinephelinae.GroupByMergingQueryRunnerV2.access$700(GroupByMergingQueryRunnerV2.java:72)
	at io.druid.query.groupby.epinephelinae.GroupByMergingQueryRunnerV2$1.make(GroupByMergingQueryRunnerV2.java:268)
	at io.druid.query.groupby.epinephelinae.GroupByMergingQueryRunnerV2$1.make(GroupByMergingQueryRunnerV2.java:155)
	at io.druid.java.util.common.guava.BaseSequence.accumulate(BaseSequence.java:43)
	at io.druid.common.guava.CombiningSequence.accumulate(CombiningSequence.java:64)
	at io.druid.java.util.common.guava.MappedSequence.accumulate(MappedSequence.java:43)
	at io.druid.java.util.common.guava.FilteredSequence.accumulate(FilteredSequence.java:45)
	at io.druid.query.groupby.epinephelinae.GroupByRowProcessor$2.make(GroupByRowProcessor.java:166)
	at io.druid.query.groupby.epinephelinae.GroupByRowProcessor$2.make(GroupByRowProcessor.java:126)
	at io.druid.java.util.common.guava.BaseSequence.accumulate(BaseSequence.java:43)
	at io.druid.common.guava.CombiningSequence.accumulate(CombiningSequence.java:64)
	at io.druid.java.util.common.guava.MappedSequence.accumulate(MappedSequence.java:43)
	at io.druid.java.util.common.guava.FilteredSequence.accumulate(FilteredSequence.java:45)
	at io.druid.query.groupby.epinephelinae.GroupByRowProcessor$2.make(GroupByRowProcessor.java:166)
	at io.druid.query.groupby.epinephelinae.GroupByRowProcessor$2.make(GroupByRowProcessor.java:126)
	at io.druid.java.util.common.guava.BaseSequence.accumulate(BaseSequence.java:43)
	at io.druid.common.guava.CombiningSequence.accumulate(CombiningSequence.java:64)
	at io.druid.java.util.common.guava.MappedSequence.accumulate(MappedSequence.java:43)
	at io.druid.java.util.common.guava.FilteredSequence.accumulate(FilteredSequence.java:45)
	at io.druid.query.groupby.epinephelinae.GroupByRowProcessor$2.make(GroupByRowProcessor.java:166)
	at io.druid.query.groupby.epinephelinae.GroupByRowProcessor$2.make(GroupByRowProcessor.java:126)
	at io.druid.java.util.common.guava.BaseSequence.accumulate(BaseSequence.java:43)
	at io.druid.common.guava.CombiningSequence.accumulate(CombiningSequence.java:64)
	at io.druid.java.util.common.guava.MappedSequence.accumulate(MappedSequence.java:43)
	at io.druid.java.util.common.guava.WrappingSequence$1.get(WrappingSequence.java:50)
	at io.druid.java.util.common.guava.SequenceWrapper.wrap(SequenceWrapper.java:55)
	at io.druid.java.util.common.guava.WrappingSequence.accumulate(WrappingSequence.java:45)
	at io.druid.java.util.common.guava.MappedSequence.accumulate(MappedSequence.java:43)
	at io.druid.java.util.common.guava.Sequences.toList(Sequences.java:150)
	at io.druid.query.groupby.GroupByQueryRunnerTestHelper.runQuery(GroupByQueryRunnerTestHelper.java:55)
	at io.druid.query.groupby.GroupByQueryMergeBufferTest.testTripleNestedGroupBy(GroupByQueryMergeBufferTest.java:355)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runners.Suite.runChild(Suite.java:127)
	at org.junit.runners.Suite.runChild(Suite.java:26)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
Caused by: java.util.concurrent.TimeoutException
	at io.druid.query.groupby.epinephelinae.GroupByMergingQueryRunnerV2.waitForFutureCompletion(GroupByMergingQueryRunnerV2.java:317)
	... 64 more
```

--- a/processing/src/test/java/io/druid/query/groupby/GroupByQueryMergeBufferTest.java
+++ b/processing/src/test/java/io/druid/query/groupby/GroupByQueryMergeBufferTest.java
@@ -59,6 +59,7 @@ import static org.junit.Assert.assertEquals;
 @RunWith(Parameterized.class)
 public class GroupByQueryMergeBufferTest
 {
+  private static final long TIMEOUT = 5000;
   private static class TestBlockingPool extends BlockingPool<ByteBuffer>
   {
     private int minRemainBufferNum;
@@ -230,7 +231,7 @@ public class GroupByQueryMergeBufferTest
         .setGranularity(Granularities.ALL)
         .setInterval(QueryRunnerTestHelper.firstToThird)
         .setAggregatorSpecs(Lists.<AggregatorFactory>newArrayList(new LongSumAggregatorFactory("rows", "rows")))
-        .setContext(ImmutableMap.of(QueryContexts.TIMEOUT_KEY, 500))
+        .setContext(ImmutableMap.of(QueryContexts.TIMEOUT_KEY, TIMEOUT))
         .build();
 
     GroupByQueryRunnerTestHelper.runQuery(factory, runner, query);
@@ -258,7 +259,7 @@ public class GroupByQueryMergeBufferTest
         .setGranularity(Granularities.ALL)
         .setInterval(QueryRunnerTestHelper.firstToThird)
         .setAggregatorSpecs(Lists.<AggregatorFactory>newArrayList(new LongSumAggregatorFactory("rows", "rows")))
-        .setContext(ImmutableMap.of(QueryContexts.TIMEOUT_KEY, 500))
+        .setContext(ImmutableMap.of(QueryContexts.TIMEOUT_KEY, TIMEOUT))
         .build();
 
     GroupByQueryRunnerTestHelper.runQuery(factory, runner, query);
@@ -297,7 +298,7 @@ public class GroupByQueryMergeBufferTest
         .setGranularity(Granularities.ALL)
         .setInterval(QueryRunnerTestHelper.firstToThird)
         .setAggregatorSpecs(Lists.<AggregatorFactory>newArrayList(new LongSumAggregatorFactory("rows", "rows")))
-        .setContext(ImmutableMap.of(QueryContexts.TIMEOUT_KEY, 500))
+        .setContext(ImmutableMap.of(QueryContexts.TIMEOUT_KEY, TIMEOUT))
         .build();
 
     GroupByQueryRunnerTestHelper.runQuery(factory, runner, query);
@@ -349,7 +350,7 @@ public class GroupByQueryMergeBufferTest
         .setGranularity(Granularities.ALL)
         .setInterval(QueryRunnerTestHelper.firstToThird)
         .setAggregatorSpecs(Lists.<AggregatorFactory>newArrayList(new LongSumAggregatorFactory("rows", "rows")))
-        .setContext(ImmutableMap.of(QueryContexts.TIMEOUT_KEY, 500))
+        .setContext(ImmutableMap.of(QueryContexts.TIMEOUT_KEY, TIMEOUT))
         .build();
 
     GroupByQueryRunnerTestHelper.runQuery(factory, runner, query);

--- a/server/src/test/java/io/druid/segment/realtime/appenderator/AppenderatorDriverTest.java
+++ b/server/src/test/java/io/druid/segment/realtime/appenderator/AppenderatorDriverTest.java
@@ -71,7 +71,7 @@ public class AppenderatorDriverTest
   private static final ObjectMapper OBJECT_MAPPER = new DefaultObjectMapper();
   private static final int MAX_ROWS_IN_MEMORY = 100;
   private static final int MAX_ROWS_PER_SEGMENT = 3;
-  private static final long PUBLISH_TIMEOUT = 5000;
+  private static final long PUBLISH_TIMEOUT = 10000;
   private static final long HANDOFF_CONDITION_TIMEOUT = 1000;
 
   private static final List<InputRow> ROWS = Arrays.<InputRow>asList(
@@ -221,13 +221,12 @@ public class AppenderatorDriverTest
       committerSupplier.setMetadata(1);
       Assert.assertTrue(driver.add(ROWS.get(0), "dummy", committerSupplier).isOk());
 
-      final SegmentsAndMetadata published = driver.publish(
+      final SegmentsAndMetadata segmentsAndMetadata = driver.publishAndRegisterHandoff(
           makeOkPublisher(),
           committerSupplier.get(),
           ImmutableList.of("dummy")
       ).get(PUBLISH_TIMEOUT, TimeUnit.MILLISECONDS);
-      final SegmentsAndMetadata segmentsAndMetadata = driver.registerHandoff(published)
-                                                            .get(HANDOFF_CONDITION_TIMEOUT, TimeUnit.MILLISECONDS);
+
       Assert.assertEquals(
           ImmutableSet.of(
               new SegmentIdentifier(DATA_SOURCE, new Interval("2000/PT1H"), VERSION, new NumberedShardSpec(0, 0))
@@ -243,13 +242,12 @@ public class AppenderatorDriverTest
       committerSupplier.setMetadata(i + 1);
       Assert.assertTrue(driver.add(ROWS.get(i), "dummy", committerSupplier).isOk());
 
-      final SegmentsAndMetadata published = driver.publish(
+      final SegmentsAndMetadata segmentsAndMetadata = driver.publishAndRegisterHandoff(
           makeOkPublisher(),
           committerSupplier.get(),
           ImmutableList.of("dummy")
       ).get(PUBLISH_TIMEOUT, TimeUnit.MILLISECONDS);
-      final SegmentsAndMetadata segmentsAndMetadata = driver.registerHandoff(published)
-                                                            .get(HANDOFF_CONDITION_TIMEOUT, TimeUnit.MILLISECONDS);
+
       Assert.assertEquals(
           ImmutableSet.of(
               // The second and third rows have the same dataSource, interval, and version, but different shardSpec of
@@ -265,13 +263,11 @@ public class AppenderatorDriverTest
     driver.persist(committerSupplier.get());
 
     // There is no remaining rows in the driver, and thus the result must be empty
-    final SegmentsAndMetadata published = driver.publish(
+    final SegmentsAndMetadata segmentsAndMetadata = driver.publishAndRegisterHandoff(
         makeOkPublisher(),
         committerSupplier.get(),
         ImmutableList.of("dummy")
     ).get(PUBLISH_TIMEOUT, TimeUnit.MILLISECONDS);
-    final SegmentsAndMetadata segmentsAndMetadata = driver.registerHandoff(published)
-                                                          .get(HANDOFF_CONDITION_TIMEOUT, TimeUnit.MILLISECONDS);
 
     Assert.assertEquals(
         ImmutableSet.of(),

